R/stack.R

```diff
@@ -391,3 +402,187 @@ sys_frame <- function(n) {
     sys.frame(n)
   }
 }
+
+#' Find the position of a frame on the evaluation stack.
+#'
+#' The position is the number of frames on the evaluation stack
+#' (\code{\link{eval_stack}()}), counting from the global frame (whose
+#' position is zero). See also \code{\link{frame_distance}()} for a
+#' function that counts frames from the opposite side (the current
+#' context).
+#'
+#' While this function returns the position of the frame on the
+#' evaluation stack, it can safely be called with intervening frames
+#' as those will be discarded.
+#'
+#' @param frame The environment of a frame. Can be any object with a
+#'   \code{\link{env}()} method. Note that for frame objects, this
+#'   simply returns \code{frame$pos}.
+#' @seealso \code{\link{frame_distance}()}
+#' @export
+#' @examples
+#' fn <- function() g(environment())
+#' g <- function(env) frame_position(env)
+#'
+#' # frame_position() returns the position of the frame on the evaluation stack:
+#' fn()
+#' identity(identity(fn()))
+#'
+#' # Note that it trims off intervening calls before counting so you
+#' # can safely nest it within other calls:
+#' g <- function(env) identity(identity(frame_position(env)))
+#' fn()
+frame_position <- function(frame) {
+  if (is_frame(frame)) {
+    return(frame$pos)
+  }
+  frame <- env(frame)
+
+  stack <- stack_trim(eval_stack(), n = 2)
+  envs <- pluck(stack, "env")
+
+  i <- 1
+  for (env in envs) {
+    if (identical(env, frame)) {
+      return(length(envs) - i)
+    }
+    i <- i + 1
+  }
+
+  stop("this environment is not on the stack", call. = FALSE)
+}
+
+#' Compute the distance between a frame and the current context.
```

Comment:
Should this be documented with `frame_position()`? 

I also wonder if you could make the names more symmetric by using `top` and `bottom` suffixes?